pbj-core/pbj-compiler/src/main/java/com/hedera/pbj/compiler/impl/Common.java

```diff
@@ -172,14 +174,367 @@ public static String javaPrimitiveToObjectType(String primitiveFieldType) {
+					generatedCodeSoFar += (
+               		"""
+             		if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + Float.hashCode($fieldName);
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.DOUBLE) {
+					generatedCodeSoFar += (
+               		"""
+					if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + Double.hashCode($fieldName);
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.STRING) {
+					generatedCodeSoFar += (
+               		"""
+					if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + $fieldName.hashCode();
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.BYTES) {
+					generatedCodeSoFar += (
+               		"""
+					if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + ($fieldName == null ? 0 : $fieldName.hashCode());
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				}
+				else if (f.parent() == null) { // process sub message
+				    generatedCodeSoFar += (
+               		"""
+					if ($fieldName != null && $fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + $fieldName.hashCode();
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				}
+				else {
+					throw new RuntimeException("Unexpected field type for getting HashCode - " + f.type().toString());
+				}
+			}
+		}
+		return generatedCodeSoFar.indent(DEFAULT_INDENT * 2);
+	}
+
+	/**
+	 * Get the hashcode codegen for a optional field.
+	 * @param generatedCodeSoFar The string that the codegen is generated into.
+	 * @param f The field for which to generate the hash code.
+	 * @return Updated codegen string.
+	 */
+	@NonNull
+	private static String getOptionalHashCodeGeneration(String generatedCodeSoFar, Field f) {
+		switch (f.messageType()) {
+			case "StringValue" -> generatedCodeSoFar += (
+     			"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + $fieldName.hashCode();
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "BoolValue" -> generatedCodeSoFar += (
+				"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + Boolean.hashCode($fieldName);
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "Int32Value", "UInt32Value" -> generatedCodeSoFar += (
+     			"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + Integer.hashCode($fieldName);
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "Int64Value", "UInt64Value" -> generatedCodeSoFar += (
+     			"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + Long.hashCode($fieldName);
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "FloatValue" -> generatedCodeSoFar += (
```

Comment:
Floats and doubles are risky to compare using == or !=, it should probably be something like `Math.abs($fieldName - DEFAULT.$fieldName) < 1e-6` - the exact epsilon value is to be checked with Jasper / Leemon

Reply:
but that is not comparison to a real value, likely it's comparison to zero and `equals` should be sufficiently robust for that